src/main/scala/cognite/spark/v1/AssetsHierarchyBuilder.scala

```diff
@@ -0,0 +1,230 @@
+package cognite.spark.v1
+
+import cats.effect.IO
+import cats.implicits._
+import cognite.spark.v1.SparkSchemaHelper.{fromRow, structType}
+import com.cognite.sdk.scala.common.SetValue
+import com.cognite.sdk.scala.v1.{Asset, AssetCreate, AssetUpdate, AssetsFilter, CogniteExternalId}
+import org.apache.spark.sql.types.{DataTypes, StructType}
+import org.apache.spark.sql.{DataFrame, SQLContext}
+
+final case class AssetsIngestSchema(
+    externalId: String,
+    parentExternalId: String,
+    source: Option[String],
+    name: String,
+    description: Option[String],
+    metadata: Option[Map[String, String]])
+
+object AssetsIngestSchema {
+  def toAssetCreate(a: AssetsIngestSchema): AssetCreate =
+    AssetCreate(
+      name = a.name,
+      description = a.description,
+      externalId = Some(a.externalId),
+      metadata = a.metadata,
+      source = a.source,
+      parentExternalId = Some(a.parentExternalId)
+    )
+
+  def toAssetUpdate(a: AssetsIngestSchema): AssetUpdate =
+    AssetUpdate(
+      name = Some(SetValue(a.name)),
+      description = a.description.map(SetValue(_)),
+      externalId = Some(SetValue(a.externalId)),
+      metadata = a.metadata.map(SetValue(_)),
+      source = a.source.map(SetValue(_)),
+      parentExternalId = Some(SetValue(a.parentExternalId))
+    )
+}
+
+final case class MultipleRootsException(message: String = "Tree has more than one root.")
+    extends Exception(message)
+final case class NoRootException(
+    message: String = """Tree has no root. Set parentExternalId to "" for the root Asset.""")
+    extends Exception(message)
+final case class InvalidTreeException(message: String = s"The tree is has an invalid structure.")
+    extends Exception(message)
+
+class AssetsHierarchyBuilder(config: RelationConfig)(val sqlContext: SQLContext)
+    extends CdfRelation(config, "assetshierarchybuilder") {
+
+  def build(df: DataFrame): IO[Unit] = {
+    val deleteMissingAssets = config.deleteMissingAssets
+    val batchSize = config.batchSize.getOrElse(Constants.DefaultBatchSize)
+
+    val sourceTree = df.collect.map(r => fromRow[AssetsIngestSchema](r))
+
+    val (root, children) = validateAndOrderInput(sourceTree)
+
+    for {
+      rootAndChildren <- getCdfRootAndTree(CogniteExternalId(root.externalId))
+      (cdfRoot, cdfChildren) = (rootAndChildren._1, rootAndChildren._2)
+      _ <- upsertRoot(root, cdfRoot)
+      (toDelete, toInsert, toUpdate) = nodesToDeleteInsertUpdate(root, children, cdfChildren)
+      _ <- insert(toInsert, batchSize)
+      _ <- update(toUpdate, batchSize)
+      _ <- delete(toDelete, deleteMissingAssets, batchSize)
+    } yield ()
+  }
+
+  def insert(toInsert: Seq[AssetsIngestSchema], batchSize: Int): IO[Unit] =
+    if (toInsert.nonEmpty) {
+      val assetCreatesToInsert = toInsert.map(AssetsIngestSchema.toAssetCreate).grouped(batchSize)
+      assetCreatesToInsert.toList.traverse(client.assets.create) *> IO.unit
```

Comment:
Maybe parTraverse instead?

Reply:
For consistency with `delete` and `update` we should format it the same
```
  toInsert...
    .grouped(batchSize)
    .toList
    .parTraverse(client.assets.create) *> IO.unit
```

Reply:
I used `.traverse` here since it (to my understanding) preserves order, which is necessary to avoid writing children before their parents exist.

Reply:
Hmm, so it is. Add a comment about that, please.